client/src/components/authFunc.tsx

```diff
@@ -0,0 +1,222 @@
+        name: string;
+        username: string;
+        email: string;
+        };
+    }
+}
+  
+  export async function fetchGitLabGraphQLQuery(
+    accessToken: string | null,
+    query: string,
+  ): Promise<GitLabData | null> {
+    try {
+      const response = await fetch('https://gitlab.com/api/graphql', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+          Authorization: `Bearer ${accessToken}`,
+        },
+        body: JSON.stringify({ query }),
+      });
+      if (!response.ok) {
+        throw new Error(`Error fetching GitLab GraphQL data: ${response.statusText}`);
+      }
+      const data = await response.json();
+      return data as GitLabData;
+    } catch (error) {
+      console.error(error);
+      return null;
+    }
+  }
+  
+
+export function expiresAtDate(tokenExpiresAt: number): void {
+    // Get the current time in milliseconds
+    const currentTimeMs = Date.now();
+
+    // Convert the current time to seconds
+    const currentTimeSec = Math.floor(currentTimeMs / 1000);
+
+    const expiresAtSec = currentTimeSec + tokenExpiresAt;
+    // Convert the seconds back to a Date object
+    const currentDate = new Date(expiresAtSec * 1000);
+
+    // Format the date as a string
+    const currentDateStr = currentDate.toLocaleString();
+
+    // Store results
+    localStorage.setItem('expiresIn', currentDateStr);
+}
+
+// Our timestamp needs to be in EU format and not NA. 
+function convertDate(dateString: string): string {
+    const date = new Date(dateString);
+    const day = date.getDate().toString().padStart(2, '0');
+    const month = (date.getMonth() + 1).toString().padStart(2, '0');
+    const year = date.getFullYear().toString();
+    const hours = date.getHours().toString().padStart(2, '0');
+    const minutes = date.getMinutes().toString().padStart(2, '0');
+    const seconds = date.getSeconds().toString().padStart(2, '0');
+    return `${day}/${month}/${year} ${hours}:${minutes}:${seconds}`;
+  }
+  
+  
+
+export const checkTokenValidity = (): boolean => {
+    if (localStorage.getItem('expiresIn')) {
+        const storedData = localStorage.getItem('expiresIn');
+        const dateEUFormat = convertDate(storedData as string);
+        console.log('test: ', dateEUFormat);
+        const dateObject = new Date(dateEUFormat);
+        const currentTime = new Date();
+        console.log('Current time: ', currentTime, 'Expires at: ', dateObject);
+        if (dateObject > currentTime) {
+            console.log('Token valid!');
+            return true;
+        } 
+        if (dateObject <= currentTime)
+        {
+            console.log('Token invalid!');
+            return false; 
+        }
+    }
+    return false;
+};
+
+
+export const callOncePerRender = (callback: () => void) => {
+    useEffect(() => {
+        callback();
+    }, []);
+};
+
+export const logoutClearLocalStorage = (): void => {
+    if (localStorage.getItem('expiresIn')) {
```

Comment:
There seems to be a [localforage](https://github.com/localForage/localForage) library which gives more flexibility in terms of browser storage. @DaloonOfDoom might also be interested in this library.